In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Data Gathering

In [4]:
# Reading the match data from the csv file
match = pd.read_csv(r'C:\Users\srini\Projects\IPL\matches.csv')
match.head(2)

id  season       city        date                team1  \
0   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06       Mumbai Indians   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   

   result  dl_applied                  winner  win_by_runs  win_by_wickets  \
0  normal           0     Sunrisers Hyderabad           35               0   
1  normal           0  Rising Pune Supergiant            0               7   

  player_of_match                                      venue         umpire1  \
0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
1       SPD Smith    Maharashtra Cricket Association Stadium  A Nand Kishore   

    umpire2 umpire3  
0  NJ Llong     NaN  
1    S Ravi     NaN

In [5]:
# Reading the data for each ball 
ball= pd.read_csv(r'C:\Users\srini\Projects\IPL\deliveries.csv')
ball.head(2)

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  

[2 rows x 21 columns]

# Data Wrangling

### Missing Values

In [59]:
match.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
id                 756 non-null int64
season             756 non-null int64
city               749 non-null object
date               756 non-null object
team1              756 non-null object
team2              756 non-null object
toss_winner        756 non-null object
toss_decision      756 non-null object
result             756 non-null object
dl_applied         756 non-null int64
winner             752 non-null object
win_by_runs        756 non-null int64
win_by_wickets     756 non-null int64
player_of_match    752 non-null object
venue              756 non-null object
umpire1            754 non-null object
umpire2            754 non-null object
umpire3            119 non-null object
dtypes: int64(5), object(13)
memory usage: 106.4+ KB


So there are total 756 matches with 752 decided winners. Let us see which rows do not have any decided winner.

In [63]:
match[match.isna()['winner']]

id  season       city        date                        team1  \
300    301    2011      Delhi  2011-05-21             Delhi Daredevils   
545    546    2015  Bangalore  2015-04-29  Royal Challengers Bangalore   
570    571    2015  Bangalore  2015-05-17             Delhi Daredevils   
744  11340    2019  Bengaluru    30/04/19  Royal Challengers Bangalore   

                           team2                  toss_winner toss_decision  \
300                Pune Warriors             Delhi Daredevils           bat   
545             Rajasthan Royals             Rajasthan Royals         field   
570  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
744             Rajasthan Royals             Rajasthan Royals         field   

        result  dl_applied winner  win_by_runs  win_by_wickets  \
300  no result           0    NaN            0               0   
545  no result           0    NaN            0               0   
570  no result           0    NaN            0               0   
744  no result           0    NaN            0               0   

    player_of_match                   venue          umpire1       umpire2  \
300             NaN        Feroz Shah Kotla        SS Hazare     RJ Tucker   
545             NaN   M Chinnaswamy Stadium        JD Cloete     PG Pathak   
570             NaN   M Chinnaswamy Stadium  HDPK Dharmasena  K Srinivasan   
744             NaN  M. Chinnaswamy Stadium      Nigel Llong  Ulhas Gandhe   

            umpire3  
300             NaN  
545             NaN  
570             NaN  
744  Anil Chaudhary

So these matches were probably washed from the rains and did not have any results declared.

In [64]:
ball.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 22 columns):
match_id            179078 non-null int64
inning              179078 non-null int64
batting_team        179078 non-null object
bowling_team        179078 non-null object
over                179078 non-null int64
ball                179078 non-null int64
batsman             179078 non-null object
non_striker         179078 non-null object
bowler              179078 non-null object
is_super_over       179078 non-null int64
wide_runs           179078 non-null int64
bye_runs            179078 non-null int64
legbye_runs         179078 non-null int64
noball_runs         179078 non-null int64
penalty_runs        179078 non-null int64
batsman_runs        179078 non-null int64
extra_runs          179078 non-null int64
total_runs          179078 non-null int64
player_dismissed    8834 non-null object
dismissal_kind      8834 non-null object
fielder             6448 non-null object
over_

This dataframe has missing values only for player dismissed and dismissal_kind. <br>

The data type for all the columns also seem to be okay.

### Duplicate values

In [66]:
match.duplicated().sum()

0

In [67]:
ball.duplicated().sum()

23

In [72]:
ball[ball.duplicated()].head()

match_id  inning                 batting_team  \
52178        221       1               Mumbai Indians   
162805      7946       1             Rajasthan Royals   
162806      7946       1             Rajasthan Royals   
162870      7946       1             Rajasthan Royals   
162964      7946       2  Royal Challengers Bangalore   

                       bowling_team  over  ball         batsman  \
52178              Delhi Daredevils     4     1    SR Tendulkar   
162805  Royal Challengers Bangalore     4     4       AM Rahane   
162806  Royal Challengers Bangalore     4     5       AM Rahane   
162870  Royal Challengers Bangalore    13     5     RA Tripathi   
162964             Rajasthan Royals    10     1  AB de Villiers   

          non_striker      bowler  is_super_over  ...  legbye_runs  \
52178         C Madan  PJ Sangwan              0  ...            0   
162805    RA Tripathi    UT Yadav              0  ...            0   
162806    RA Tripathi    UT Yadav              0  ...            0   
162870      AM Rahane   YS Chahal              0  ...            0   
162964  Mandeep Singh     I Sodhi              0  ...            0   

        noball_runs  penalty_runs  batsman_runs  extra_runs  total_runs  \
52178             0             0             1           0           1   
162805            0             0             4           0           4   
162806            0             0             1           0           1   
162870            0             0             0           0           0   
162964            0             0             0           0           0   

        player_dismissed  dismissal_kind fielder over_ball  
52178                NaN             NaN     NaN       4.1  
162805               NaN             NaN     NaN       4.4  
162806               NaN             NaN     NaN       4.5  
162870               NaN             NaN     NaN      13.5  
162964               NaN             NaN     NaN      10.1  

[5 rows x 22 columns]

In [73]:
ball.query('match_id==221 & inning==1 & over==4 & ball==1')

match_id  inning    batting_team      bowling_team  over  ball  \
52169       221       1  Mumbai Indians  Delhi Daredevils     4     1   
52178       221       1  Mumbai Indians  Delhi Daredevils     4     1   

            batsman non_striker      bowler  is_super_over  ...  legbye_runs  \
52169  SR Tendulkar     C Madan  PJ Sangwan              0  ...            0   
52178  SR Tendulkar     C Madan  PJ Sangwan              0  ...            0   

       noball_runs  penalty_runs  batsman_runs  extra_runs  total_runs  \
52169            0             0             1           0           1   
52178            0             0             1           0           1   

       player_dismissed  dismissal_kind fielder over_ball  
52169               NaN             NaN     NaN       4.1  
52178               NaN             NaN     NaN       4.1  

[2 rows x 22 columns]

So these duplicate values need to be removed as they dont add any value.

In [74]:
ball.drop_duplicates(inplace= True)

### Creating summary dataframes from ball dataframe

In [75]:
# Creating a new df summarizing total runs and wickets that fell in each innings

summary= ball.groupby(by=['match_id','inning','batting_team']).sum()['total_runs'].reset_index()
temp = ball.groupby(by=['match_id','inning','batting_team']).count()['player_dismissed'].reset_index()

summary= pd.merge(summary, temp,how='outer', on= ['match_id','inning', 'batting_team'])

In [76]:
# Creating a new column combining the over and the ball bowled
ball['over_ball']= ball['over']+ball['ball']/10

In [77]:
# Merging the total balls used in the innings with the summary data
temp= ball.groupby(by=['match_id','inning','batting_team']).max()['over_ball'].reset_index()

summary= pd.merge(summary, temp,how='outer', on= ['match_id','inning', 'batting_team'])
summary.head(2)

match_id  inning                 batting_team  total_runs  \
0         1       1          Sunrisers Hyderabad         207   
1         1       2  Royal Challengers Bangalore         172   

   player_dismissed  over_ball  
0                 4       20.6  
1                10       20.4

### Creating a batsman summary dataframe

In [78]:
# Total runs scored by each batsman
batsman= ball.groupby(by='batsman').sum()['batsman_runs'].reset_index()

#Total balls played by each batsman
temp = ball.groupby(by='batsman').count()['ball']
batsman= pd.merge(batsman, temp, how='outer',on='batsman')

# Total 4s hit
temp = ball.query('batsman_runs==4').groupby(by='batsman').count()['batsman_runs'].reset_index()
batsman = pd.merge(batsman, temp, how='outer',on='batsman')

#Total 6s hit
temp= ball.query('batsman_runs==6').groupby(by='batsman').count()['batsman_runs'].reset_index()
batsman = pd.merge(batsman, temp, how='outer',on='batsman')

batsman.rename(columns={'batsman_runs_x':'total_runs','batsman_runs_y':'4s', 'batsman_runs':'6s'}, inplace= True)
batsman.head(3)

batsman  total_runs  ball    4s    6s
0  A Ashish Reddy         280   196  16.0  15.0
1      A Chandila           4     7   NaN   NaN
2        A Chopra          53    75   7.0   NaN

In [157]:
# Calculating the strike rate
batsman['strike_rate']= (batsman.total_runs/ batsman.ball)*100
batsman.head(2)

batsman  total_runs  ball    4s    6s  strike_rate
0  A Ashish Reddy         280   196  16.0  15.0   142.857143
1      A Chandila           4     7   NaN   NaN    57.142857

In [161]:
# Calculating percent of the runs scored by boundries
batsman['boundary_percent']= (batsman['4s']*4+ batsman['6s']*6)/ batsman.total_runs
batsman.head(1)

batsman  total_runs  ball    4s    6s  strike_rate  boundary_percent
0  A Ashish Reddy         280   196  16.0  15.0   142.857143              0.55

### Creating a bowler summary dataframe

In [86]:
ball.dismissal_kind.value_counts()

caught                   5348
bowled                   1581
run out                   852
lbw                       540
stumped                   278
caught and bowled         211
retired hurt               12
hit wicket                 10
obstructing the field       2
Name: dismissal_kind, dtype: int64

For calculating the wickets picked by the bowler we need to exclude the run outs, retired hurt and obstructing the field. There are actually 2 dismissals for obstructing the field!!! <br>

A quick Google search reveals that they were probable runouts but the batsmen (Yusuf Pathan and Amit Mishra) blocked the ball from hitting the stumps and hence were declared out.

In [111]:
ball.head(2)

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  legbye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...            0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...            0   

   noball_runs  penalty_runs  batsman_runs  extra_runs  total_runs  \
0            0             0             0           0           0   
1            0             0             0           0           0   

   player_dismissed  dismissal_kind fielder over_ball  
0               NaN             NaN     NaN       1.1  
1               NaN             NaN     NaN       1.2  

[2 rows x 22 columns]

In [173]:
# Total balls bowled and total runs given by the bowler
bowler= ball.groupby(by='bowler').count()['ball'].reset_index()
temp= ball.groupby(by='bowler').sum()['total_runs']

bowler= pd.merge(bowler, temp, on='bowler', how='outer')

In [174]:
# Filtering the runouts/ retired hurts
temp = ball.query('dismissal_kind!="obstructing the field" & dismissal_kind !="retired hurt" & dismissal_kind != "run out"')

# Filtering the wickets taken
temp = temp[~temp.isna().dismissal_kind][['bowler','dismissal_kind']]
temp= temp.groupby(by=['bowler']).count().reset_index()

bowler= pd.merge(bowler, temp, on='bowler', how='outer')

In [175]:
# Renaming the column to reflect the wickets
bowler.rename(columns={'dismissal_kind':'wickets_taken'}, inplace= True)

In [176]:
# Calculating the total dot balls
temp = ball.query('total_runs==0').groupby(by='bowler').count()['ball'].reset_index()
bowler= pd.merge(bowler, temp, on='bowler', how='outer')

In [177]:
# Renaming the column to reflect the wickets
bowler.rename(columns={'ball_x':'balls_bowled', 'ball_y':'dot_balls'}, inplace= True)
bowler.head(1)

bowler  balls_bowled  total_runs  wickets_taken  dot_balls
0  A Ashish Reddy           270         400           18.0       79.0

In [178]:
# Calculating the number of bowled and LBWs taken by the bowler
temp =ball.query('dismissal_kind=="bowled" | dismissal_kind =="lbw" ').groupby(by='bowler').count()['dismissal_kind'].reset_index()
bowler= pd.merge(bowler, temp, on='bowler', how='outer')
bowler.head(3)

bowler  balls_bowled  total_runs  wickets_taken  dot_balls  \
0  A Ashish Reddy           270         400           18.0       79.0   
1      A Chandila           234         245           11.0      103.0   
2     A Choudhary           108         144            5.0       42.0   

   dismissal_kind  
0             9.0  
1             NaN  
2             NaN

In [179]:
# Renaming the column to reflect the wickets
bowler.rename(columns={'dismissal_kind':'bowl_or_lbw'}, inplace= True)

In [180]:
bowler.head(1)

bowler  balls_bowled  total_runs  wickets_taken  dot_balls  \
0  A Ashish Reddy           270         400           18.0       79.0   

   bowl_or_lbw  
0          9.0

In [181]:
# Calculating the bowler economy
bowler['economy']= bowler.total_runs/ (bowler.balls_bowled/6)

In [182]:
# Calculating dot ball percent
bowler['dot_ball_percent']= bowler.dot_balls/bowler.balls_bowled

In [183]:
# Calculating the runs scored in last 3 overs (death over) to compare the bowlers performance
a= ball.query('over>=18').groupby('bowler').sum()['total_runs']
b= ball.query('over>=18').groupby('bowler').count()['batsman_runs']
temp =pd.DataFrame(data=[a,b] ).transpose().reset_index()
temp.rename(columns={'total_runs':'death_runs','batsman_runs':'death_balls' }, inplace= True)

bowler= pd.merge(bowler, temp, on='bowler', how='outer')
bowler.head(3)

bowler  balls_bowled  total_runs  wickets_taken  dot_balls  \
0  A Ashish Reddy           270         400           18.0       79.0   
1      A Chandila           234         245           11.0      103.0   
2     A Choudhary           108         144            5.0       42.0   

   bowl_or_lbw   economy  dot_ball_percent  death_runs  death_balls  
0          9.0  8.888889          0.292593        73.0         38.0  
1          NaN  6.282051          0.440171         NaN          NaN  
2          NaN  8.000000          0.388889        29.0         21.0

In [187]:
# Calculating bowling economy in death over
bowler['death_econ']= bowler.death_runs/ (bowler.death_balls/6)

In [189]:
# Calculating the percentage of bowled and LBWs by the bowler among the wickets they took
bowler['bowl_lbw_percent']= bowler.bowl_or_lbw/ bowler.wickets_taken

In [190]:
bowler.head(1)

bowler  balls_bowled  total_runs  wickets_taken  dot_balls  \
0  A Ashish Reddy           270         400           18.0       79.0   

   bowl_or_lbw   economy  dot_ball_percent  death_runs  death_balls  \
0          9.0  8.888889          0.292593        73.0         38.0   

   death_econ  bowl_lbw_percent  
0   11.526316               0.5

In [184]:

#ball.groupby(by=['batting_team','inning','over']).sum()['total_runs'].head(10)